### ISSUE TITLE
[Phase 2] Reject Inference with EM Algorithm

### ISSUE THREAD

**Issue by Sakeeb91** [Sakeeb91/credit-scoring-reject-inference#3]
## Summary

Implement reject inference using the Expectation-Maximization (EM) algorithm to incorporate rejected applicants into model training. This addresses the selection bias problem where only approved applicants have observable default outcomes.

**Part of:** #1 (Meta Issue)
**Depends on:** #2 (Phase 1: WoE Encoding)

## System Context

```
src/reject_inference/
├── __init__.py
├── em_algorithm.py      # Core EM implementation (THIS PHASE)
└── pseudo_labeling.py   # Confidence-based labels (THIS PHASE)

tests/
└── test_reject_inference.py  # Unit tests (THIS PHASE)
```

### The Selection Bias Problem

Traditional credit models are trained only on approved applicants:
- **Approved + Defaulted**: Observable
- **Approved + Non-default**: Observable
- **Rejected**: Outcome unknown (missing labels)

Ignoring rejects biases the model because rejected applicants likely have higher default risk than the approved population.

### EM Algorithm Approach

1. **Initialize**: Train model on approved-only data
2. **E-step**: Predict P(default) for rejected applicants (soft labels)
3. **M-step**: Retrain model using approved + rejected with soft labels
4. **Iterate**: Until convergence (log-likelihood stabilizes)

## Objectives

- [ ] Implement EM algorithm for binary classification with missing labels
- [ ] Add pseudo-labeling with configurable confidence thresholds
- [ ] Implement convergence monitoring with early stopping
- [ ] Multiple random restarts to avoid local optima
- [ ] Visualization of inferred label distributions

### Code Skeleton

📄 **File:** `src/reject_inference/em_algorithm.py`

```python
import numpy as np
from sklearn.linear_model import LogisticRegression
from typing import Tuple, Optional, List


class EMRejectInference:
    """EM algorithm for reject inference in credit scoring."""

    def __init__(self, max_iter: int = 100, tol: float = 1e-4,
                 n_restarts: int = 5, random_state: int = 42):
        self.max_iter = max_iter
        self.tol = tol
        self.n_restarts = n_restarts
        self.random_state = random_state
        self.model_: Optional[LogisticRegression] = None
        self.log_likelihoods_: List[float] = []
        self.rejected_probs_: Optional[np.ndarray] = None

    def fit(self, X_approved: np.ndarray, y_approved: np.ndarray,
            X_rejected: np.ndarray) -> 'EMRejectInference':
        """
        Fit model using EM to incorporate rejected applicants.

        Args:
            X_approved: Features for approved applicants (n_approved, n_features)
            y_approved: Binary outcomes for approved (0=non-default, 1=default)
            X_rejected: Features for rejected applicants (n_rejected, n_features)

        Returns:
            self
        """
        best_ll = -np.inf
        best_model = None

        for restart in range(self.n_restarts):
            model, ll_history = self._run_em(
                X_approved, y_approved, X_rejected,
                random_state=self.random_state + restart
            )

            if ll_history[-1] > best_ll:
                best_ll = ll_history[-1]
                best_model = model
                self.log_likelihoods_ = ll_history

        self.model_ = best_model
        self.rejected_probs_ = self.model_.predict_proba(X_rejected)[:, 1]
        return self

    def _run_em(self, X_approved, y_approved, X_rejected,
                random_state) -> Tuple[LogisticRegression, List[float]]:
        """Single EM run."""
        # Initialize with approved-only model
        model = LogisticRegression(random_state=random_state, max_iter=1000)
        model.fit(X_approved, y_approved)

        ll_history = []
        n_approved = len(y_approved)
        n_rejected = len(X_rejected)

        for iteration in range(self.max_iter):
            # E-step: Compute expected labels for rejected
            rejected_probs = model.predict_proba(X_rejected)[:, 1]

            # M-step: Refit with soft labels
            X_combined = np.vstack([X_approved, X_rejected])

            # For rejected, create weighted samples
            # Each rejected sample contributes twice: once as default, once as non-default
            # weighted by their predicted probabilities
            sample_weights = np.concatenate([
                np.ones(n_approved),
                rejected_probs,  # Weight for default class
                1 - rejected_probs  # Weight for non-default class
            ])

            y_combined = np.concatenate([
                y_approved,
                np.ones(n_rejected),  # Default labels for rejected
                np.zeros(n_rejected)  # Non-default labels for rejected
            ])

            X_expanded = np.vstack([X_approved, X_rejected, X_rejected])

            model = LogisticRegression(random_state=random_state, max_iter=1000)
            model.fit(X_expanded, y_combined, sample_weight=sample_weights)

            # Compute log-likelihood
            ll = self._compute_log_likelihood(
                X_approved, y_approved, X_rejected, model
            )
            ll_history.append(ll)

            # Check convergence
            if len(ll_history) > 1:
                if abs(ll_history[-1] - ll_history[-2]) < self.tol:
                    break

        return model, ll_history

    def _compute_log_likelihood(self, X_approved, y_approved,
                                 X_rejected, model) -> float:
        """Compute observed data log-likelihood."""
        # Approved: standard log-likelihood
        probs_approved = model.predict_proba(X_approved)
        ll_approved = np.sum(
            y_approved * np.log(probs_approved[:, 1] + 1e-10) +
            (1 - y_approved) * np.log(probs_approved[:, 0] + 1e-10)
        )

        # Rejected: marginal likelihood (sum over both outcomes)
        probs_rejected = model.predict_proba(X_rejected)
        ll_rejected = np.sum(np.log(
            probs_rejected[:, 0] + probs_rejected[:, 1] + 1e-10
        ))

        return ll_approved + ll_rejected

    def predict_proba(self, X: np.ndarray) -> np.ndarray:
        """Predict probabilities using fitted model."""
        return self.model_.predict_proba(X)
```

### Pseudo-Labeling Alternative

📄 **File:** `src/reject_inference/pseudo_labeling.py`

```python
import numpy as np
from sklearn.base import BaseEstimator
from typing import Optional


class PseudoLabelInference:
    """
    Pseudo-labeling for reject inference.

    Only uses rejected samples where model is confident.
    """

    def __init__(self, confidence_threshold: float = 0.8,
                 max_iterations: int = 5):
        self.confidence_threshold = confidence_threshold
        self.max_iterations = max_iterations
        self.model_: Optional[BaseEstimator] = None

    def fit(self, X_approved: np.ndarray, y_approved: np.ndarray,
            X_rejected: np.ndarray,
            base_model: BaseEstimator) -> 'PseudoLabelInference':
        """
        Iteratively add confident pseudo-labels.
        """
        X_train = X_approved.copy()
        y_train = y_approved.copy()
        X_unlabeled = X_rejected.copy()

        for iteration in range(self.max_iterations):
            # Fit model
            self.model_ = base_model.fit(X_train, y_train)

            if len(X_unlabeled) == 0:
                break

            # Predict on unlabeled
            probs = self.model_.predict_proba(X_unlabeled)
            confidence = np.max(probs, axis=1)
            predictions = self.model_.predict(X_unlabeled)

            # Select confident samples
            confident_mask = confidence >= self.confidence_threshold

            if confident_mask.sum() == 0:
                break

            # Add to training set
            X_train = np.vstack([X_train, X_unlabeled[confident_mask]])
            y_train = np.concatenate([y_train, predictions[confident_mask]])

            # Remove from unlabeled
            X_unlabeled = X_unlabeled[~confident_mask]

        return self
```

### Debugging Scenarios

| Issue | Symptom | Solution |
|-------|---------|----------|
| EM doesn't converge | Log-likelihood oscillates | Increase `max_iter`, check data quality |
| All rejects same class | `rejected_probs_` all near 0 or 1 | Lower confidence threshold, check feature overlap |
| Numerical instability | NaN in log-likelihood | Add small epsilon (1e-10) to log arguments |
| Local optima | Different restarts give different results | Increase `n_restarts`, use ensemble |

### Validation Approach

Create synthetic data with known ground truth for rejects:
1. Generate full population with labels
2. Simulate approval process (keep harder cases)
3. Compare inferred labels to true labels
4. Measure AUC improvement over approve-only model

## Definition of Done

- [ ] EM algorithm converges on synthetic data
- [ ] Log-likelihood increases monotonically
- [ ] Multiple restarts implemented
- [ ] Pseudo-labeling alternative available
- [ ] Convergence visualization generated
- [ ] Tests pass with coverage > 85%

## Files to Modify

| File | Lines | Action | Description |
|------|-------|--------|-------------|
| `src/reject_inference/em_algorithm.py` | 1-120 | Create | Core EM implementation |
| `src/reject_inference/pseudo_labeling.py` | 1-60 | Create | Pseudo-labeling alternative |
| `tests/test_reject_inference.py` | 1-80 | Create | Unit tests |

## Resources

- [EM Algorithm Tutorial](https://www.cs.utah.edu/~piyush/teaching/EM_algorithm.pdf)
- [Reject Inference in Credit Scoring](https://www.sciencedirect.com/science/article/abs/pii/S0377221703007707)
